include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -250,78 +260,163 @@ __parallel_for(_ExecutionPolicy&& __exec, _Fp __brick, _Index __count, _Ranges&&
+        __iters_per_work_item = __ceiling_div(__n, (__max_compute_units * __work_group_size));
+        _PRINT_INFO_IN_DEBUG_MODE(__exec, __work_group_size, __max_compute_units);
+    }
 
     ::std::size_t __size_per_work_group =
-        __iters_per_work_item * __work_group_size;            // number of buffer elements processed within workgroup
-    _Size __n_groups = (__n - 1) / __size_per_work_group + 1; // number of work groups
-    _Size __n_items = (__n - 1) / __iters_per_work_item + 1;  // number of work items
-
-    _PRINT_INFO_IN_DEBUG_MODE(__exec, __work_group_size, __max_compute_units);
+        __iters_per_work_item * __work_group_size; // number of buffer elements processed within workgroup
+    ::std::size_t __n_groups = __ceiling_div(__n, __size_per_work_group); // number of work groups
+    ::std::size_t __n_items = __ceiling_div(__n, __iters_per_work_item);  // number of work items
 
     // Create temporary global buffers to store temporary values
     sycl::buffer<_Tp> __temp(sycl::range<1>(2 * __n_groups));
+    sycl::buffer<_Tp> __res(sycl::range<1>(1));
     // __is_first == true. Reduce over each work_group
     // __is_first == false. Reduce between work groups
     bool __is_first = true;
 
-    // For memory utilization it's better to use one big buffer instead of two small because size of the buffer is close to a few MB
-    _Size __offset_1 = 0;
-    _Size __offset_2 = __n_groups;
+    // For memory utilization it's better to use one big buffer instead of two small because size of the buffer is close
+    // to a few MB
+    ::std::size_t __offset_1 = 0;
+    ::std::size_t __offset_2 = __n_groups;
 
     sycl::event __reduce_event;
     do
     {
-        __reduce_event = __exec.queue().submit([&, __is_first, __offset_1, __offset_2, __n, __n_items,
-                                                __n_groups](sycl::handler& __cgh) {
+        __reduce_event = __exec.queue().submit([&, __is_first, __offset_1, __offset_2, __n, __n_items, __n_groups,
+                                                __iters_per_work_item](sycl::handler& __cgh) {
             __cgh.depends_on(__reduce_event);
 
-            oneapi::dpl::__ranges::__require_access(__cgh, __rngs...); //get an access to data under SYCL buffer
+            oneapi::dpl::__ranges::__require_access(__cgh, __rngs...); // get an access to data under SYCL buffer
             auto __temp_acc = __temp.template get_access<access_mode::read_write>(__cgh);
+            auto __res_acc = __res.template get_access<access_mode::write>(__cgh);
             __dpl_sycl::__local_accessor<_Tp> __temp_local(sycl::range<1>(__work_group_size), __cgh);
 #if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
             __cgh.use_kernel_bundle(__kernel.get_kernel_bundle());
 #endif
-            __cgh.parallel_for<_ReduceKernel>(
+            __cgh.parallel_for<__i_kernel_name<_ReduceKernel, _IsGPU{}>>(
```

Comment:
Since this is recursive with more than one iteration possible (at least on GPU), does it make sense to use the single group   scan algorithm as the leaf node rather than continuing to recurse?

Reply:
I like the idea and will prototype it. This seems especially beneficial to me for mid-sized input arrays. Very large arrays might benefit from the exposed concurrency of the recursive tree reduction though.